auth/jwt/middleware.go

```diff
@@ -0,0 +1,107 @@
+package jwt
+
+import (
+	"errors"
+
+	"golang.org/x/net/context"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/go-kit/kit/endpoint"
+)
+
+const (
+	// JWTTokenContextKey holds the key used to store a JWT Token in the context
+	JWTTokenContextKey = "JWTToken"
+	// JWTClaimsContxtKey holds the key used to store the JWT Claims in the context
+	JWTClaimsContextKey = "JWTClaims"
+)
+
+var (
+	ErrTokenContextMissing     = errors.New("Token up for parsing was not passed through the context")
+	ErrTokenInvalid            = errors.New("JWT Token was invalid")
+	ErrUnexpectedSigningMethod = errors.New("Unexptected signing method")
+	ErrKIDNotFound             = errors.New("Key ID was not found in key set")
+	ErrNoKIDHeader             = errors.New("Token doesn't have 'kid' header")
+)
+
+type Claims map[string]interface{}
+
+type KeySet map[string]struct {
+	Method jwt.SigningMethod
+	Key    []byte
+}
+
+// Create a new JWT token generating middleware, specifying signing method and the claims
+// you would like it to contain. Particularly useful for clients.
+func NewSigner(kid string, keys KeySet, claims Claims) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, request interface{}) (response interface{}, err error) {
+			key, ok := keys[kid]
+			if !ok {
+				return nil, ErrKIDNotFound
+			}
+
+			token := jwt.NewWithClaims(key.Method, jwt.MapClaims(claims))
+			token.Header["kid"] = kid
+			// Sign and get the complete encoded token as a string using the secret
+			tokenString, err := token.SignedString(key.Key)
+			if err != nil {
+				return nil, err
+			}
+			ctx = context.WithValue(ctx, JWTTokenContextKey, tokenString)
+
+			return next(ctx, request)
+		}
+	}
+}
+
+// Create a new JWT token parsing middleware, specifying a jwt.Keyfunc interface and the
+// signing method. Adds the resulting claims to endpoint context or returns error on invalid
+// token. Particularly useful for servers.
+func NewParser(keys KeySet) endpoint.Middleware {
```

Comment:
I could be wrong, as I have never used auth0 or stormpath, but my understanding is those services implement single sign on authentication for users to authenticate to an API. As I designed it (and again I could have misinterpreted the intention) this package implements service-to-service authentication. Meaning we are authenticating that a request is coming from another service inside our perimeter. These requests would be accompanied with JWT tokens that are signed inside a client from our own infrastructure (somehow sharing secret keys between the two services). This would allow us to pass a KID header with the shared Key ID and, if this is the correct intention of this package, then handling user authenticated tokens from a SSO provider would be additional functionality built into the service code.